pyiron_atomistics/atomistics/structure/neighbors.py

```diff
@@ -79,26 +107,63 @@ def __repr__(self):
             + "- distances : Distances to the neighbors of given positions\n"
             + "- indices : Indices of the neighbors of given positions\n"
         )
-        if self.vecs is not None:
+        if self._vecs is not None:
             to_return += "- vecs : Vectors to the neighbors of given positions\n"
         return to_return
 
     def copy(self):
         new_neigh = Tree(self._ref_structure)
-        new_neigh.distances = self.distances.copy()
-        new_neigh.vecs = self.vecs.copy()
-        new_neigh.indices = self.indices.copy()
+        new_neigh._distances = self._distances.copy()
+        new_neigh._indices = self._indices.copy()
         new_neigh._extended_positions = self._extended_positions
         new_neigh._wrapped_indices = self._wrapped_indices
-        new_neigh._allow_ragged = self._allow_ragged
         new_neigh._extended_indices = self._extended_indices
         new_neigh.wrap_positions = self.wrap_positions
         new_neigh._tree = self._tree
         new_neigh.num_neighbors = self.num_neighbors
         new_neigh.cutoff_radius = self.cutoff_radius
         new_neigh._norm_order = self._norm_order
+        new_neigh._positions = self._positions.copy()
         return new_neigh
 
+    def _reshape(self, value, key=None, ref_vector=None):
+        if key is None:
+            key = self.mode
+        if key == 'filled':
+            return value
+        elif key == 'ragged':
+            return self._contract(value, ref_vector=ref_vector)
+        elif key == 'flattened':
+            return value[self._distances < np.inf]
+
+    @property
+    def distances(self):
+        """Distances to neighboring atoms"""
+        return self._reshape(self._distances)
+
+    @property
+    def vecs(self):
+        """Vectors to neighboring atoms"""
+        if self._vecs is None:
+            self._vecs = self._get_vectors(
+                positions=self._positions,
+                distances=self._distances,
+                indices=self._extended_indices
+            )
+        return self._reshape(self._vecs)
+
+    @property
+    def indices(self):
+        """Indices of neighboring atoms"""
+        return self._reshape(self._indices)
+
+    @property
+    def atom_numbers(self):
```

Comment:
I didn't get a chance to mention it in the meeting, but I find this variable name unintuitive. I feel like it should somehow also utilize the root "index" to make its relation to the `indices` attribute clear, but I struggle to think of something I really like.

In the same way that we make `volume` and `energy_X` singular in MD output, what do you think about making this singular to contrast it to the plural `indices` where there's more than one value in each row? Something like `target_index`?

Reply:
Yeah I've been having the same problem for years and I never came up with a good name. So far I'm glad that @sudarsan-surendralal suggested `atom_numbers`, but it's true that the relation to `indices` might still be somewhat unclear.

Since we tend to use plural for per-atom quantities, I might actually prefer to keep plural here as well. I like `target_index`, but it still might be difficult to know whether the atoms themselves are meant or the neighbours. What do you think of `self_indices`?